### PR TITLE
REGR: Fix assignment bug for unary operators

### PR DESCRIPTION
**Pull request by dsaxton**
- [x] closes https://github.com/pandas-dev/pandas/issues/39943
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

If there was somehow a way to copy the mask only on write so we aren't always copying that may be better. @simonjayhawkins I put this in the `IntegerArray` tests even though we're technically checking a `Series`, let me know if you'd prefer I rewrite or move elsewhere.

### CODE REVIEW

**Comment by simonjayhawkins** [comment]
> If there was somehow a way to copy the mask only on write so we aren't always copying that may be better. 

for a backport, let's keep this simple.

File: pandas/tests/arrays/integer/test_arithmetic.py
Comment (simonjayhawkins) [inline pandas/tests/arrays/integer/test_arithmetic.py:322]: can we also test \_\_invert__, since I believe has the same issue.

Comment (dsaxton) [reply pandas/tests/arrays/integer/test_arithmetic.py]: Good catch, that required another change

**Comment by jreback** [comment]
> If there was somehow a way to copy the mask only on write so we aren't always copying that may be better. @simonjayhawkins I put this in the IntegerArray tests even though we're technically checking a Series, let me know if you'd prefer I rewrite or move elsewhere.

this is not necessary, these copies are *very* cheap,

File: pandas/tests/arrays/integer/test_arithmetic.py
Comment (jreback) [inline pandas/tests/arrays/integer/test_arithmetic.py:322]: can you test for boolean as well

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/arrays/integer/test_arithmetic.py
Comment (dsaxton) [reply pandas/tests/arrays/integer/test_arithmetic.py:322]: Sure, added

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. waiting on CI.

**Comment by simonjayhawkins** [comment]
```
2021-02-23T01:07:24.4100333Z 
2021-02-23T01:07:24.4100753Z ================================== FAILURES ===================================
2021-02-23T01:07:24.4101532Z _________________________ TestAccessor.test_from_coo __________________________
2021-02-23T01:07:24.4102065Z [gw0] win32 -- Python 3.7.9 C:\Miniconda\envs\pandas-dev\python.exe
2021-02-23T01:07:24.4102361Z 
2021-02-23T01:07:24.4103135Z self = <pandas.tests.arrays.sparse.test_array.TestAccessor object at 0x000001317E6BC888>
2021-02-23T01:07:24.4103462Z 
2021-02-23T01:07:24.4103797Z     @td.skip_if_no_scipy
2021-02-23T01:07:24.4104207Z     def test_from_coo(self):
2021-02-23T01:07:24.4104597Z         import scipy.sparse
2021-02-23T01:07:24.4104947Z     
2021-02-23T01:07:24.4105283Z         row = [0, 3, 1, 0]
2021-02-23T01:07:24.4105674Z         col = [0, 3, 1, 2]
2021-02-23T01:07:24.4106045Z         data = [4, 5, 7, 9]
2021-02-23T01:07:24.4106494Z         sp_array = scipy.sparse.coo_matrix((data, (row, col)))
2021-02-23T01:07:24.4106960Z         result = pd.Series.sparse.from_coo(sp_array)
2021-02-23T01:07:24.4107356Z     
2021-02-23T01:07:24.4107834Z         index = pd.MultiIndex.from_arrays([[0, 0, 1, 3], [0, 2, 1, 3]])
2021-02-23T01:07:24.4108389Z         expected = pd.Series([4, 9, 7, 5], index=index, dtype="Sparse[int]")
2021-02-23T01:07:24.4109198Z >       tm.assert_series_equal(result, expected)
2021-02-23T01:07:24.4109817Z E       AssertionError: Attributes of Series are different
2021-02-23T01:07:24.4110213Z E       
2021-02-23T01:07:24.4110548Z E       Attribute "dtype" are different
2021-02-23T01:07:24.4110965Z E       [left]:  Sparse[float64, nan]
2021-02-23T01:07:24.4111366Z E       [right]: Sparse[int32, 0]
2021-02-23T01:07:24.4111579Z 
2021-02-23T01:07:24.4112141Z pandas\tests\arrays\sparse\test_array.py:1193: AssertionError
2021-02-23T01:07:24.4112871Z ============================== warnings summary ===============================
```

**Comment by simonjayhawkins** [comment]
will merge after the #39979 is merged. (otherwise may not be able to auto backport)

**Comment by simonjayhawkins** [comment]
@meeseeksdev backport 1.2.x

File: doc/source/whatsnew/v1.2.3.rst
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v1.2.3.rst:19]: For future reference, I think we should generally say something like "nullable integer dtype" instead of "IntegerArray", as most users should never deal / know about IntegerArray

File: pandas/core/arrays/integer.py
Comment (jorisvandenbossche) [inline pandas/core/arrays/integer.py:322]: We should do the same for `__pos__`?

File: pandas/tests/arrays/masked/test_arithmetic.py
Comment (jorisvandenbossche) [inline pandas/tests/arrays/masked/test_arithmetic.py:169]: you can use the `data` fixture here as used in the other tests above

File: pandas/core/arrays/integer.py
Comment (simonjayhawkins) [reply pandas/core/arrays/integer.py:322]: probably not for a backport. but indeed this needs some discussion to ensure consistency across the codebase. see TimedeltaArray

```
    def __pos__(self) -> TimedeltaArray:
        return type(self)(self._data, freq=self.freq)
```

```
>>> import pandas as pd
>>> pd.__version__
'1.3.0.dev0+804.g3289f82975'
>>>
>>> arr = pd.timedelta_range(0, periods=10)._values
>>> arr
<TimedeltaArray>
['0 days', '1 days', '2 days', '3 days', '4 days', '5 days', '6 days',
 '7 days', '8 days', '9 days']
Length: 10, dtype: timedelta64[ns]
>>>
>>> arr2 = +arr
>>>
>>> arr2 is arr
False
>>>
>>> arr2._data is arr._data
True
>>>
>>> arr2[5] = None
>>>
>>> arr
<TimedeltaArray>
['0 days', '1 days', '2 days', '3 days', '4 days',      NaT, '6 days',
 '7 days', '8 days', '9 days']
Length: 10, dtype: timedelta64[ns]
>>>
>>>
```

Comment (jorisvandenbossche) [reply pandas/core/arrays/integer.py:322]: OK, I see, so that's indeed something we should fix more generally. Is there already an issue about it?

Comment (dsaxton) [reply pandas/core/arrays/integer.py:322]: In the context here I had been thinking that something like `s1 = +s` should behave the same way as `s1 = s`, but actually I think you are right. Likely `+s` should return something new instead of simply a reference to the thing.

File: doc/source/whatsnew/v1.2.3.rst
Comment (dsaxton) [reply doc/source/whatsnew/v1.2.3.rst:19]: Valid point, I can update this

File: pandas/core/arrays/integer.py
Comment (simonjayhawkins) [reply pandas/core/arrays/integer.py:322]: > Is there already an issue about it?

no. will need to create one, since the issue that this PR addresses is slightly different. #39943 is about different arrays sharing a mask, whereas returning self for \_\_pos__, although maybe incorrect, does not create the inconsistencies when assigning null and non-null vales

File: pandas/tests/arrays/masked/test_arithmetic.py
Comment (simonjayhawkins) [reply pandas/tests/arrays/masked/test_arithmetic.py:169]: this is being addressed in #39916, so no action required here

File: doc/source/whatsnew/v1.2.3.rst
Comment (simonjayhawkins) [reply doc/source/whatsnew/v1.2.3.rst:19]: done. #40019. thanks @dsaxton

**Pull request merged**
